### PR TITLE
Advanced login - remember the user

### LINKED ISSUE #12: Authentication
#### As an application hoster I want to add user authentication to my application.

__Story__:

We have done a great job providing immediate value to our customers. At this
point they can both add and delete images, in addition to a few other features
that are either completed or in the works. Along with a growth in our web
service's popularity have come an increase in number of reports of missing
images. Upon investigation, it turns out that the majority of delete requests
are issued by only a handful of IP addresses. While we could blacklist those IP
addresses from the delete operation, it has always been our plan to only permit
the user who added an image to delete that image. To begin that effort we must
first add the concept of a user.

Please split this task up into three separate PRs. The second, and third should
depend on the preceding ones.

__Acceptance criteria PR1__ (reviewer Ozhe):
- [x] Users can sign up.

__Acceptance criteria PR2__ (reviewer bboe):
- [ ] Users can log in after they've signed up.
- [ ] Users can log out after logging in.
- [ ] Logged in users should not be able to register, nor log in again.

__Acceptance criteria PR3__ (reviewer bboe):
- [ ] Users can choose to be remembered as part of the login
  process.<sup>†</sup>
- [ ] Multiple browsers can concurrently maintain the remembered status for the
  same user.

<sup>†</sup>: "Remembering" means that users will remain logged into the site
even after completely closing their web browser. Please be sure to completely
close the browser (`cmd` + `q`) when testing this functionality. Closing the
window (`cmd` + `w`, or clicking the red circle) on Mac OS does not close the
browser.

__Note__: Do not use Devise for this task.

__Discussion topics__:
- [ ] If we wanted to add more advanced features such as email activation and
  "forgot my password", should we continue to "roll our own" authentication?

__Reference material__:
- https://www.railstutorial.org/book/modeling_users
- https://www.railstutorial.org/book/sign_up
- https://www.railstutorial.org/book/basic_login
- https://www.railstutorial.org/book/advanced_login

__Further reading__:
- https://github.com/plataformatec/devise

__Dependencies__:
- Delete Image
- Refactor flow tests to use page objects

### PR DESCRIPTION
**Pull request by butzhang**
Closes #12 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -18,4 +19,8 @@ def destroy
   def fetch_user
     @user = User.find_by(email: params[:session][:email].downcase)
   end
+
+  def check_remember_me?
+    params[:session][:remember_me] == '1' ? true : false
```
Comment (bboe) [inline app/controllers/sessions_controller.rb]: Is the ternary necessary?

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -3,6 +3,7 @@ class SessionsController < ApplicationController
   def create
     if @user && @user.authenticate(params[:session][:password])
       log_in @user
+      check_remember_me? ? remember(@user) : forget(@user)
```
Comment (bboe) [inline app/controllers/sessions_controller.rb]: Does `remember` and `forget` exist at this point?

File: test/integration/users_login_test.rb [hunk 3]
```diff
@@ -27,4 +27,24 @@ class UsersLoginTest < ActionDispatch::IntegrationTest
     assert_select 'a[href=?]', logout_path,      count: 0
     assert_select 'a[href=?]', user_path(@user), count: 0
   end
+
+  test 'login with remembering ' do
+    log_in_as(@user, remember_me: '1')
+    assert_predicate cookies['remember_token'], :present?
+    assert_equal cookies['remember_token'], assigns(:user).remember_token
```
Comment (bboe) [inline test/integration/users_login_test.rb]: Can you avoid using `assigns`?

File: test/integration/users_login_test.rb [hunk 4]
```diff
@@ -27,4 +27,24 @@ class UsersLoginTest < ActionDispatch::IntegrationTest
...
+  end
+
+  test 'login without remembering' do
```
Comment (bboe) [inline test/integration/users_login_test.rb]: Great test.

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -14,4 +14,8 @@ def destroy
     log_out
     redirect_to root_path
   end
+
+  def fetch_user
+    @user = User.find_by(email: params[:session][:email].downcase)
```
Comment (bboe) [inline app/controllers/sessions_controller.rb]: Why have `@user` as an instance variable?

File: app/helpers/sessions_helper.rb [hunk 6]
```diff
@@ -4,7 +4,16 @@ def log_in(user)
   end
 
   def current_user
-    @current_user ||= User.find_by(id: session[:user_id])
+    return @current_user if @current_user
+    if (user_id = session[:user_id])
+      @current_user ||= User.find_by(id: user_id)
```
Comment (bboe) [inline app/helpers/sessions_helper.rb]: Is there ever a case where `||=` is necessary here?

File: app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -4,7 +4,16 @@ def log_in(user)
...
+    elsif (user_id = cookies.signed[:user_id])
+      user = User.find_by(id: user_id)
+      if user && user.authenticated?(cookies[:remember_token])
```
Comment (bboe) [inline app/helpers/sessions_helper.rb]: What's the purpose of `authenticated?`?

File: test/integration/users_login_test.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
+require 'test_helper'
+
+class UsersLoginTest < ActionDispatch::IntegrationTest
+  setup do
+    @user = User.create!(name: 'Michael Example',
+                         email: 'michael@example.com',
+                         password: 'password')
+  end
+  test 'login with valid information followed by logout' do
+    get login_path
```
Comment (bboe) [inline test/integration/users_login_test.rb]: What's the point of `get login_path` here?

File: test/integration/users_login_test.rb [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    post login_path, params: { session: { email:    @user.email,
+                                          password: 'password' } }
+    assert_predicate session[:user_id], :present?
+    assert_redirected_to @user
+    follow_redirect!
+    assert_select 'a[href=?]', login_path, count: 0
+    assert_select 'a[href=?]', logout_path
+    assert_select 'a[href=?]', user_path(@user)
+    delete logout_path
```
Comment (bboe) [inline test/integration/users_login_test.rb]: Should logging out be its own test case?

File: test/integration/users_login_test.rb (hunk 8 above)
Comment (butzhang) [inline test/integration/users_login_test.rb]: yeah, it should be removed

File: test/integration/users_login_test.rb (hunk 9 above)
Comment (bboe) [inline test/integration/users_login_test.rb]: Looks like this still isn't handled.

Comment (butzhang) [inline test/integration/users_login_test.rb]: Do you imply that I should pull out the test `logout after login` and put it in its own test class, like `UserLogoutTest`?

**Pull request merged**